src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -3693,11 +3693,57 @@ void MacroAssembler::load_method_holder(Register holder, Register method) {
 }
 
 void MacroAssembler::load_klass(Register dst, Register src) {
-  if (UseCompressedClassPointers) {
-    ldrw(dst, Address(src, oopDesc::klass_offset_in_bytes()));
-    decode_klass_not_null(dst);
-  } else {
-    ldr(dst, Address(src, oopDesc::klass_offset_in_bytes()));
+  assert(UseCompressedClassPointers, "expects UseCompressedClassPointers");
+
+  // We can receive src and dst in the same register here, and they can also
+  // come in rscratch1 and rscratch2. Let's allocate an additional register
+  // here to preserve src across the fast-path.
+  Register tmp = dst;
+  if (src == dst) {
+    if (src == r0) {
+      tmp = r1;
+    } else {
+      tmp = r0;
+    }
+    push(RegSet::of(tmp), sp);
+  }
+  assert_different_registers(src, tmp);
+  assert_different_registers(src, rscratch1);
+  assert_different_registers(src, rscratch2);
+
+  Label slow, done;
+
+  // Check if we can take the (common) fast path, if obj is unlocked.
+  ldr(tmp, Address(src, oopDesc::mark_offset_in_bytes()));
+  eor(tmp, tmp, markWord::unlocked_value);
+  tst(tmp, markWord::lock_mask_in_place);
+  br(Assembler::NE, slow);
```

Comment:
```suggestion
  andr(dst, tmp, markWord::lock_mask_in_place);
  cbnz(dst, Assembler::NE, slow);
```

Reply:
I believe that wouldn't work either. The fast path that is following this code requires that the upper 32 bits (the narrowKlass) of tmp remain intact: it shifts them to the right by 32bits and decodes the narrowKlass.

Reply:
> I believe that wouldn't work either. The fast path that is following this code requires that the upper 32 bits (the narrowKlass) of tmp remain intact: it shifts them to the right by 32bits and decodes the narrowKlass.

So what? tmp is untouched.

Reply:
> > I believe that wouldn't work either. The fast path that is following this code requires that the upper 32 bits (the narrowKlass) of tmp remain intact: it shifts them to the right by 32bits and decodes the narrowKlass.
> 
> So what? tmp is untouched.

We don't guarantee that tmp != dst. We only guarantee that tmp != src to avoid clobbering src.

Reply:
OK, I give up. Please document 'clobbers flags' very clearly.

Reply:
> OK, I give up. Please document 'clobbers flags' very clearly.

I will try to make it that the caller ensures that src != dst, and also passes a tmp register, then we can avoid clobbering the flags register altogether and also avoid the register shuffling madness. Not sure if that's always feasable, though.

Reply:
Alright, ensuring src != dst is relatively easy and seems worth doing (makes fast-path faster). I tried using rscratch2 as tmp register and use the proposed andr/cbnz sequence, but apparently rscratch2 is live in at least one path, and I could not easily find it. I documented the register usage, and that the load_klass() clobbers condition flags. That seems good enough for now, especially considering that at some point in the future, it will all go away (when locking is not overlaying the Klass* stuff in the header anymore).
@theRealAph ok now?